### PR TITLE
OAuth spec alignment

### PR DESCRIPTION
**Pull request by matthieusieben**
This PR fixes several details in order to make atproto's reference implementation match the atproto Oauth spec.

### CODE REVIEW

File: .changeset/late-wolves-itch.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@atproto/oauth-provider": patch
+---
+
+Enfore code_challenge_method=S256 request parameter
```
Comment (devinivy) [inline .changeset/late-wolves-itch.md:5]:
```suggestion
Enforce code_challenge_method=S256 request parameter
```

File: packages/oauth/oauth-provider/src/errors/invalid-client-metadata-error.ts [hunk 2]
```diff
@@ -12,8 +12,16 @@ export class InvalidClientMetadataError extends OAuthError {
...
+  static from(
+    cause: unknown,
+    fallbackMessage = 'Invalid client metadata document',
+  ): OAuthError {
+    if (cause instanceof OAuthError) {
+      return cause
+    }
+    if (cause instanceof TypeError) {
+      return new InvalidClientMetadataError(cause.message, cause)
+    }
```
Comment (devinivy) [inline packages/oauth/oauth-provider/src/errors/invalid-client-metadata-error.ts:24]: I don't think we need to change this, but it is a bit interesting that a system `TypeError` (e.g. "null has no properties") could make it into a user-facing error, if I'm understanding correctly.

Comment (matthieusieben) [reply packages/oauth/oauth-provider/src/errors/invalid-client-metadata-error.ts:24]: This is due to the fact that this error is meant to be used in the context of parsing a client_id, meaning that any TypeError (e.g. InvalidURL) is actually not a code error but rather a user data error. I could move that logic elsewhere where `TypeError` are actually thrown ?

Comment (matthieusieben) [reply packages/oauth/oauth-provider/src/errors/invalid-client-metadata-error.ts:24]: I added a comment to explain the reasonning.

**Review by Vincentroland — APPROVED** [review]

**Review by Vincentroland — APPROVED** [review]

**Comment by matthieusieben** [comment]
Likely to fix  #2814

File: .changeset/great-news-shout.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@atproto/oauth-types": patch
+---
+
+Enfoce ClientID URL path to be normalized
```
[the remark this answers is not in the document]
Comment (devinivy) [reply .changeset/great-news-shout.md:5]: I took a closer look, and I see now that the path normalization check is fairly in line with draft-parecki-oauth-client-id-metadata-document-01 👍

File: packages/oauth/oauth-types/src/oauth-client-id-loopback.ts [hunk 4]
```diff
@@ -15,44 +18,89 @@ export function isOAuthClientIdLoopback<C extends OAuthClientId>(
...
+
+  if (clientId.length === queryStringIdx) {
+    return {} // no query string to parse
   }
 
-  if (url.hash) {
-    throw new TypeError('Loopback ClientID must not contain a fragment')
+  if (clientId[queryStringIdx] !== '?') {
+    throw new TypeError('Loopback ClientID must not contain a path component')
   }
```
Comment (devinivy) [inline packages/oauth/oauth-types/src/oauth-client-id-loopback.ts:53]: Do we handle the case that there's no path component and also no query params?

**Review by devinivy — APPROVED** [review]

File: packages/oauth/oauth-types/src/oauth-client-id-loopback.ts (hunk 4 above)
Comment (matthieusieben) [reply packages/oauth/oauth-types/src/oauth-client-id-loopback.ts:53]: Yep, in the previous check (`if (clientId.length === queryStringIdx)`)

**Pull request merged**